internal/provider/helpers.go

```diff
@@ -0,0 +1,256 @@
+		opts.GitUsername = data.GitUsername.ValueString()
+	}
+
+	if !data.GitPassword.IsNull() {
+		overrides["ENVBUILDER_GIT_PASSWORD"] = struct{}{}
+		opts.GitPassword = data.GitPassword.ValueString()
+	}
+
+	if !data.IgnorePaths.IsNull() {
+		overrides["ENVBUILDER_IGNORE_PATHS"] = struct{}{}
+		opts.IgnorePaths = tfutil.TFListToStringSlice(data.IgnorePaths)
+	}
+
+	if !data.Insecure.IsNull() {
+		overrides["ENVBUILDER_INSECURE"] = struct{}{}
+		opts.Insecure = data.Insecure.ValueBool()
+	}
+
+	if data.RemoteRepoBuildMode.IsNull() {
+		opts.RemoteRepoBuildMode = true
+	} else {
+		overrides["ENVBUILDER_REMOTE_REPO_BUILD_MODE"] = struct{}{}
+		opts.RemoteRepoBuildMode = data.RemoteRepoBuildMode.ValueBool()
+	}
+
+	if !data.SSLCertBase64.IsNull() {
+		overrides["ENVBUILDER_SSL_CERT_BASE64"] = struct{}{}
+		opts.SSLCertBase64 = data.SSLCertBase64.ValueString()
+	}
+
+	if !data.Verbose.IsNull() {
+		overrides["ENVBUILDER_VERBOSE"] = struct{}{}
+		opts.Verbose = data.Verbose.ValueBool()
+	}
+
+	if !data.WorkspaceFolder.IsNull() {
+		overrides["ENVBUILDER_WORKSPACE_FOLDER"] = struct{}{}
+		opts.WorkspaceFolder = data.WorkspaceFolder.ValueString()
+	}
+
+	// convert extraEnv to a map for ease of use.
+	extraEnv := make(map[string]string)
+	for k, v := range data.ExtraEnv.Elements() {
+		extraEnv[k] = tfutil.TFValueToString(v)
+	}
+	diags = append(diags, overrideOptionsFromExtraEnv(&opts, extraEnv, overrides)...)
+
+	return opts, diags
+}
+
+// overrideOptionsFromExtraEnv overrides the options in opts with values from extraEnv.
+// It returns any diagnostics encountered.
+// It will not override certain options, such as ENVBUILDER_CACHE_REPO and ENVBUILDER_GIT_URL.
+func overrideOptionsFromExtraEnv(opts *eboptions.Options, extraEnv map[string]string, overrides map[string]struct{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	// Make a map of the options for easy lookup.
+	optsMap := make(map[string]pflag.Value)
+	for _, opt := range opts.CLI() {
+		optsMap[opt.Env] = opt.Value
+	}
+	for key, val := range extraEnv {
+		if _, found := nonOverrideOptions[key]; found {
+			diags.AddAttributeWarning(path.Root("extra_env"),
+				"Cannot override required environment variable",
+				fmt.Sprintf("The key %q in extra_env cannot be overridden.", key),
+			)
+			continue
+		}
+
+		// Check if the option was set on the provider data model and generate a warning if so.
+		if _, overridden := overrides[key]; overridden {
+			diags.AddAttributeWarning(path.Root("extra_env"),
+				"Overriding provider environment variable",
+				fmt.Sprintf("The key %q in extra_env overrides an option set on the provider.", key),
+			)
+		}
+
+		// XXX: workaround for serpent behaviour where calling Set() on a
+		// string slice will append instead of replace: set to empty first.
+		if key == "ENVBUILDER_IGNORE_PATHS" {
+			_ = optsMap[key].Set("")
+		}
+
+		opt, found := optsMap[key]
+		if !found {
+			// ignore unknown keys
+			continue
+		}
```

Comment:
Actually, this should probably be before some of the other checks as well as everything depends on it.